### DISCUSSION TITLE
How can I update the position of a wall throughout the simulation?

### DISCUSSION THREAD

**Discussion by aldorvasq** [glotzerlab/hoomd-blue#1874] [Support] [answered]
Dear HOOMD users,

I am trying to implement a movable wall in my current simulation (HOOMD 4.4.1). As of now I am trying to implement this by using ``hoomd.custom.Action`` and I can not find the right way to use it. I am basing most of the code I use (shown below) from the example provided in ["How to continuously vary potential parameters"](https://hoomd-blue.readthedocs.io/en/v4.4.1/howto/continuously-vary-potential-parameters.html) and from a discussion post from HOOMDv3 (["Compressing Walls in HOOMD 3"](https://groups.google.com/g/hoomd-users/c/oh1enxNoJ3Y)) where it is suggested to update the walls' position using:
```python
wall_potential.walls = [# new walls as a function of the current simulation state, timestep, etc...]
``` 
I have tried to implement that, but I get a ``RuntimeError: Buffer is full`` message which I am not sure if its due to my system being ran in a newer version of HOOMD or to something else.

Here is a snippet of my attempted implementation, where I employ two walls (planes) parallel to each other separated by 70 units of distance in the x-direction. I attempt to move the left wall towards the right (+x-direction) 10 units of distance during a time span of  20,000 timesteps.

```python
left_wall = hoomd.wall.Plane(origin=(-35, 0, 0), normal=(1, 0, 0))
right_wall = hoomd.wall.Plane(origin=(35, 0, 0), normal=(-1, 0, 0))
lj_wall = hoomd.md.external.wall.ForceShiftedLJ(walls=[left_wall, right_wall])
lj_wall.params[['W']] = {"epsilon": 1.0, "sigma": 1.0, "r_cut": 2.0 ** (1 / 6)}
integrator.forces.append(lj_wall)

sim.operations.integrator = integrator
# Step 1: Subclass hoomd.custom.Action.
class WallParameterModifier(hoomd.custom.Action):
    def __init__(self, lj_wall):
        super().__init__()
        self.lj_wall = lj_wall

    def act(self, timestep):
        wall_position = -35 + timestep * 0.0005 
        left = hoomd.wall.Plane(origin=(wall_position, 0, 0), normal=(1, 0, 0))
        self.lj_wall.walls = [left, right_wall]

# Step 2: Create a hoomd.update.CustomUpdater
wall_parameter_modifier = WallParameterModifier(lj_wall)
wall_parameter_updater = hoomd.update.CustomUpdater(
    trigger=hoomd.trigger.Periodic(1), action=wall_parameter_modifier)

# Step 3: Add the updater to the operations
sim.operations.updaters.append(wall_parameter_updater)
```

I also tried different things in the ``act(self, timestep)`` function and the simulation does run, but in everything I tried the walls remain fixed in the initial positions they were assigned (``hoomd.wall.Plane(origin=(-35, 0, 0), normal=(1, 0, 0))``). Here is one of the things I tried:
```python
    def act(self, timestep):
        wall_position = -35 + timestep * 0.0005 
        left = hoomd.wall.Plane(origin=(wall_position, 0, 0), normal=(1, 0, 0))
        self.lj_wall = hoomd.md.external.wall.ForceShiftedLJ(walls=[left, right_wall])
```
**Could someone point in the right direction as to how implement a movable wall in HOOMD?** 

I will leave a front-facing image of the simulation box (80x60x60) I am using, where I left extra space in the x-direction to be able to test if the walls were working.
![Screenshot from 2024-08-28 18-23-20](https://github.com/user-attachments/assets/5d4b00b5-4b5a-47b0-8a15-7256491d43da)

Thank you,
Aldo V

**Comment by joaander** [comment, marked as answer]
Use: 
```
    def act(self, timestep):
        wall_position = -35 + timestep * 0.0005 
        left = hoomd.wall.Plane(origin=(wall_position, 0, 0), normal=(1, 0, 0))
        self.lj_wall.walls[0] = left
```

Your original script should have worked. I am able to reproduce the `Buffer is full` error message but did not see an obvious reason for the error. I have opened #1879 to track the defect.
